apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -85,53 +108,182 @@ public override int GetHashCode()
                         (Endpoint?.GetHashCode() ?? 0);
                 }
             }
+
+            /// <inheritdoc />
+            public override string ToString()
+            {
+                return $"Credential:{Credential?.ToString() ?? "null"} EndPoint:{Endpoint}";
+            }
         }
 
-        private class ClientPoolEntry
+        private class CredentialClientPool
         {
-            private readonly ClientPoolKey _key;
-            private SpannerClient _client;
-            private volatile Task<SpannerClient> _creationTask;
+            private readonly ClientCredentialKey _key;
+            private readonly PriorityList<SpannerClientCreator> _clientPriorityList =
+                new PriorityList<SpannerClientCreator>();
             private readonly object _sync = new object();
 
-            public ClientPoolEntry(ClientPoolKey key) => _key = key;
 
-            public async Task<SpannerClient> AcquireClientFromEntryAsync()
+            public CredentialClientPool(ClientCredentialKey key) => _key = key;
+
+            //Returns the total sum of reference counts.
+            internal int DumpCredentialPoolContents(StringBuilder stringBuilder)
+            {
+                lock (_sync)
+                {
+                    int referenceCountTotal = 0;
+                    int i = 0;
+                    foreach (var item in _clientPriorityList.GetSnapshot())
+                    {
+                        stringBuilder.AppendLine($"  {i}:{item}");
+                        referenceCountTotal += item.RefCount;
+                        i++;
+                    }
+                    return referenceCountTotal;
+                }
+            }
+
+            public Task<SpannerClient> AcquireClientAsync()
             {
-                if (_client == null)
+                Task<SpannerClient> result;
+
+                lock (_sync)
                 {
-                    Logger.Debug(() => "Creating a new SpannerClient.");
-                    var endpoint = _key.Endpoint ?? SpannerClient.DefaultEndpoint;
-                    if (_key.Credential != null)
+                    //first ensure that the pool is of the correct size.
+                    while (_clientPriorityList.Count > SpannerOptions.Instance.MaximumGrpcChannels)
```

Comment:
Is it worth grabbing a local copy of SpannerOptions.Instance.MaximumGrpcChannels and using that, so we know we'll hit a single number rather than potentially being messed up by it changing over time? It would be a very odd situation, but it might be easier to reason about.